src/bijectors/planar_layer.jl

```diff
@@ -23,74 +23,110 @@ function Base.:(==)(b1::PlanarLayer, b2::PlanarLayer)
     return PlanarLayer(w, u, b)
 end
 
-planar_flow_m(x) = -1 + softplus(x)   # for planar flow from A.1
-ψ(z, w, b) = (1 .- tanh.(w' * z .+ b).^2) .* w    # for planar flow from eq(11)
+"""
+    get_u_hat(u::AbstractVector{<:Real}, w::AbstractVector{<:Real})
 
-# An internal version of transform that returns intermediate variables
-function _transform(flow::PlanarLayer, z::AbstractVecOrMat)
-    return _planar_transform(flow.u, flow.w, first(flow.b), z)
-end
-function _planar_transform(u, w, b, z)
-    u_hat = get_u_hat(u, w)
-    transformed = z .+ u_hat .* tanh.(w' * z .+ b) # from eq(10)
-    return (transformed = transformed, u_hat = u_hat)
-end
+Return a tuple of vector ``û`` that guarantees invertibility of the planar layer, and
+scalar ``wᵀ û``.
 
-(b::PlanarLayer)(z) = _transform(b, z).transformed
+# Mathematical background
 
-function forward(flow::PlanarLayer, z::AbstractVecOrMat)
-    transformed, u_hat = _transform(flow, z)
-    # Compute log_det_jacobian
-    psi = ψ(z, flow.w, first(flow.b)) .+ zero(eltype(u_hat))
-    if psi isa AbstractVector
-        T = eltype(psi)
-    else
-        T = typeof(vec(psi))
-    end
-    log_det_jacobian::T = log.(abs.(1 .+ psi' * u_hat)) # from eq(12)
-    return (rv = transformed, logabsdetjac = log_det_jacobian)
+According to appendix A.1, vector ``û`` defined by
+```math
+û(w, u) = u + (\\log(1 + \\exp{(wᵀu)}) - 1 - wᵀu) \\frac{w}{\\|w\\|²}
+```
+guarantees that the planar layer ``f(z) = z + û tanh(wᵀz + b)`` is invertible.
+We can rewrite ``û`` as
+```math
+û = u + (\\log(1 + \\exp{(-wᵀu)}) - 1) \\frac{w}{\\|w\\|²}.
+```
+
+Additionally, we obtain
+```math
+wᵀû = wᵀu + \\log(1 + \\exp{(-wᵀu)}) - 1 = \\log(1 + \\exp{(wᵀu)}) - 1.
+```
+
+# References
+
+D. Rezende, S. Mohamed (2015): Variational Inference with Normalizing Flows.
+arXiv:1505.05770
+"""
+function get_u_hat(u::AbstractVector{<:Real}, w::AbstractVector{<:Real})
+    wT_u = dot(w, u)
+    û = u .+ ((softplus(-wT_u) - 1) / sum(abs2, w)) .* w
+    wT_û = softplus(wT_u) - 1
+    return û, wT_û
 end
 
-function (ib::Inverse{<:PlanarLayer})(y::AbstractVector{<:Real})
-    flow = ib.orig
+# Helper functions
+aT_b(a::AbstractVector{<:Real}, b::AbstractVector{<:Real}) = dot(a, b)
+aT_b(a::AbstractVector{<:Real}, b::AbstractMatrix{<:Real}) = permutedims(a) * b
+
+_vec(x::AbstractArray{<:Real}) = vec(x)
+_vec(x::Real) = x
+
+# An internal version of the transform in eq. (10) that returns intermediate variables
+function _transform(flow::PlanarLayer, z::AbstractVecOrMat{<:Real})
     w = flow.w
     b = first(flow.b)
-    u_hat = get_u_hat(flow.u, w)
+    û, wT_û = get_u_hat(flow.u, w)
+    wT_z = aT_b(w, z)
+    transformed = z .+ û .* tanh.(wT_z .+ b)
+    return (transformed = transformed, wT_û = wT_û, wT_z = wT_z)
+end
+
+(b::PlanarLayer)(z) = _transform(b, z).transformed
 
-    # Find the scalar ``alpha`` from A.1.
-    wt_y = dot(w, y)
-    wt_u_hat = dot(w, u_hat)
-    alpha = find_alpha(wt_y, wt_u_hat, b)
+#=
```

Comment:
Is there a reason why this isn't ``` rather than `#=`?

Reply:
It felt wrong to add a docstring to the generic `forward` method, I thought a regular comment might be better here.

Reply:
Ah, nono good point! Just leave it:)